File: ios/MockDeviceManager.swift
#if DEBUG
import Foundation
import MWDATCore
import MWDATMockDevice

/// Manages mock devices for testing without physical hardware.
/// Only available in DEBUG builds.
@MainActor
public final class MockDeviceManager {
    public static let shared = MockDeviceManager()

    private let logger = EMWDATLogger.shared

    /// Map of deviceIdentifier → MockRaybanMeta instance
    private var devices: [String: any MockRaybanMeta] = [:]

    private init() {}

    // MARK: - Kit Lifecycle

    public func enableMockDeviceKit(initiallyRegistered: Bool = true, initialPermissionsGranted: Bool = true) {
        let config = MockDeviceKitConfig(
            initiallyRegistered: initiallyRegistered,
            initialPermissionsGranted: initialPermissionsGranted
        )
        MockDeviceKit.shared.enable(config: config)
        logger.info("MockDeviceManager", "MockDeviceKit enabled", context: [
            "initiallyRegistered": initiallyRegistered,
            "initialPermissionsGranted": initialPermissionsGranted
        ])
    }

    public func disableMockDeviceKit() {
        MockDeviceKit.shared.disable()
        devices.removeAll()
        logger.info("MockDeviceManager", "MockDeviceKit disabled")
    }

    public func isMockDeviceKitEnabled() -> Bool {
        return MockDeviceKit.shared.isEnabled
    }

    // MARK: - Pair / Unpair

    public func pairMockDevice() -> String {
        let device = MockDeviceKit.shared.pairRaybanMeta()
        let id = "\(device.deviceIdentifier)"
        devices[id] = device
        logger.info("MockDeviceManager", "Paired mock device", context: ["id": id])
        return id
    }

    public func unpairMockDevice(id: String) throws {
        guard let device = devices[id] else {
            throw MockDeviceManagerError.deviceNotFound(id)
        }
        MockDeviceKit.shared.unpairDevice(device)
        devices.removeValue(forKey: id)
        logger.info("MockDeviceManager", "Unpaired mock device", context: ["id": id])
    }

    public func getMockDevices() -> [String] {
        return Array(devices.keys)
    }

    // MARK: - Power

    public func powerOn(id: String) throws {
        try getDevice(id).powerOn()
    }

    public func powerOff(id: String) throws {
        try getDevice(id).powerOff()
    }

    // MARK: - Don / Doff

    public func don(id: String) throws {
        try getDevice(id).don()
    }

    public func doff(id: String) throws {
        try getDevice(id).doff()
    }

    // MARK: - Fold / Unfold

    public func fold(id: String) throws {
        try getDevice(id).fold()
    }

    public func unfold(id: String) throws {
        try getDevice(id).unfold()
    }

    // MARK: - Camera (file-based — now synchronous in SDK 0.6)

    public func setCameraFeed(id: String, fileURL: URL) throws {
        let camera = try getDevice(id).services.camera
        camera.setCameraFeed(fileURL: fileURL)
    }

    public func setCapturedImage(id: String, fileURL: URL) throws {
        let camera = try getDevice(id).services.camera
        camera.setCapturedImage(fileURL: fileURL)
    }

    // MARK: - Camera (phone camera — new in SDK 0.6)

    public func setCameraFeedFromCamera(id: String, facing: String) async throws {
        let camera = try getDevice(id).services.camera
        let cameraFacing: CameraFacing = facing == "back" ? .back : .front
        await camera.setCameraFeed(cameraFacing: cameraFacing)
    }

    // MARK: - Permissions

    public func setPermissionStatus(permission: String, status: String) {
        guard let perm = mapPermission(permission),
              let stat = mapPermissionStatus(status) else { return }
        MockDeviceKit.shared.permissions.set(perm, stat)
        logger.info("MockDeviceManager", "Set permission status", context: [
            "permission": permission,
            "status": status
        ])
    }

    public func setPermissionRequestResult(permission: String, result: String) {
        guard let perm = mapPermission(permission),
              let stat = mapPermissionStatus(result) else { return }
        MockDeviceKit.shared.permissions.setRequestResult(perm, result: stat)
        logger.info("MockDeviceManager", "Set permission request result", context: [
            "permission": permission,
            "result": result
        ])
    }

    // MARK: - Helpers

    private func getDevice(_ id: String) throws -> any MockRaybanMeta {
        guard let device = devices[id] else {
            throw MockDeviceManagerError.deviceNotFound(id)
        }
        return device
    }

    private func mapPermission(_ permission: String) -> Permission? {
        switch permission {
        case "camera": return .camera
        default: return nil
        }
    }

    private func mapPermissionStatus(_ status: String) -> PermissionStatus? {
        switch status {
        case "granted": return .granted
        case "denied": return .denied
        default: return nil
        }
    }
}

public enum MockDeviceManagerError: LocalizedError {
    case deviceNotFound(String)

    public var errorDescription: String? {
        switch self {
        case .deviceNotFound(let id):
            return "Mock device not found: \(id)"
        }
    }
}
#endif

File: android/src/main/java/expo/modules/emwdat/MockDeviceManager.kt
package expo.modules.emwdat

import android.content.Context
import android.net.Uri
import com.meta.wearable.dat.core.types.Permission
import com.meta.wearable.dat.core.types.PermissionStatus
import com.meta.wearable.dat.mockdevice.MockDeviceKit
import com.meta.wearable.dat.mockdevice.api.MockDeviceKitConfig
import com.meta.wearable.dat.mockdevice.api.MockDeviceKitInterface
import com.meta.wearable.dat.mockdevice.api.MockRaybanMeta
import com.meta.wearable.dat.mockdevice.api.camera.CameraFacing

object MockDeviceManager {
    private val logger = EMWDATLogger
    private var devices: MutableMap<String, MockRaybanMeta> = mutableMapOf()
    private var mockDeviceKit: MockDeviceKitInterface? = null

    private fun getKit(context: Context): MockDeviceKitInterface {
        return mockDeviceKit ?: MockDeviceKit.getInstance(context.applicationContext).also {
            mockDeviceKit = it
        }
    }

    // MARK: - Kit Lifecycle

    fun enableMockDeviceKit(context: Context, initiallyRegistered: Boolean, initialPermissionsGranted: Boolean) {
        val kit = getKit(context)
        val config = MockDeviceKitConfig(
            initiallyRegistered = initiallyRegistered,
            initialPermissionsGranted = initialPermissionsGranted
        )
        kit.enable(config)
        logger.info("MockDeviceManager", "MockDeviceKit enabled", mapOf(
            "initiallyRegistered" to initiallyRegistered,
            "initialPermissionsGranted" to initialPermissionsGranted
        ))
    }

    fun disableMockDeviceKit(context: Context) {
        val kit = getKit(context)
        kit.disable()
        devices.clear()
        logger.info("MockDeviceManager", "MockDeviceKit disabled")
    }

    fun isMockDeviceKitEnabled(context: Context): Boolean {
        val kit = getKit(context)
        return kit.isEnabled
    }

    // MARK: - Pair / Unpair

    fun pairMockDevice(context: Context): String {
        val kit = getKit(context)
        val device = kit.pairRaybanMeta()
        val id = device.deviceIdentifier.toString()
        devices[id] = device
        logger.info("MockDeviceManager", "Paired mock device", mapOf("id" to id))
        return id
    }

    fun unpairMockDevice(context: Context, id: String) {
        val device = devices[id] ?: throw IllegalArgumentException("Mock device not found: $id")
        val kit = getKit(context)
        kit.unpairDevice(device)
        devices.remove(id)
        logger.info("MockDeviceManager", "Unpaired mock device", mapOf("id" to id))
    }

    fun getMockDevices(): List<String> = devices.keys.toList()

    // MARK: - Power

    fun powerOn(id: String) {
        getDevice(id).powerOn()
    }

    fun powerOff(id: String) {
        getDevice(id).powerOff()
    }

    // MARK: - Don / Doff

    fun don(id: String) {
        getDevice(id).don()
    }

    fun doff(id: String) {
        getDevice(id).doff()
    }

    // MARK: - Fold / Unfold

    fun fold(id: String) {
        getDevice(id).fold()
    }

    fun unfold(id: String) {
        getDevice(id).unfold()
    }

    // MARK: - Camera (file-based)

    fun setCameraFeed(id: String, fileUrl: String) {
        val device = getDevice(id)
        val uri = parseUri(fileUrl)
        device.services.cameraKit.setCameraFeed(uri)
        logger.info("MockDeviceManager", "Set camera feed", mapOf("id" to id, "uri" to uri.toString()))
    }

    fun setCapturedImage(id: String, fileUrl: String) {
        val device = getDevice(id)
        val uri = parseUri(fileUrl)
        device.services.cameraKit.setCapturedImage(uri)
        logger.info("MockDeviceManager", "Set captured image", mapOf("id" to id, "uri" to uri.toString()))
    }

    // MARK: - Camera (phone camera — new in SDK 0.6)

    fun setCameraFeedFromCamera(id: String, facing: String) {
        val device = getDevice(id)
        val cameraFacing = if (facing == "back") CameraFacing.BACK else CameraFacing.FRONT
        device.services.cameraKit.setCameraFeed(cameraFacing)
        logger.info("MockDeviceManager", "Set camera feed from camera", mapOf("id" to id, "facing" to facing))
    }

    // MARK: - Permissions

    fun setPermissionStatus(context: Context, permission: String, status: String) {
        val kit = getKit(context)
        val perm = mapPermission(permission) ?: return
        val stat = mapPermissionStatus(status) ?: return
        kit.permissions.set(perm, stat)
        logger.info("MockDeviceManager", "Set permission status", mapOf(
            "permission" to permission,
            "status" to status
        ))
    }

    fun setPermissionRequestResult(context: Context, permission: String, result: String) {
        val kit = getKit(context)
        val perm = mapPermission(permission) ?: return
        val stat = mapPermissionStatus(result) ?: return
        kit.permissions.setRequestResult(perm, stat)
        logger.info("MockDeviceManager", "Set permission request result", mapOf(
            "permission" to permission,
            "result" to result
        ))
    }

    // MARK: - Helpers

    private fun getDevice(id: String): MockRaybanMeta {
        return devices[id] ?: throw IllegalArgumentException("Mock device not found: $id")
    }

    private fun parseUri(fileUrl: String): Uri {
        return if (fileUrl.startsWith("file://")) {
            Uri.parse(fileUrl)
        } else {
            Uri.parse("file://$fileUrl")
        }
    }

    private fun mapPermission(permission: String): Permission? = when (permission) {
        "camera" -> Permission.CAMERA
        else -> null
    }

    private fun mapPermissionStatus(status: String): PermissionStatus? = when (status) {
        "granted" -> PermissionStatus.Granted
        "denied" -> PermissionStatus.Denied
        else -> null
    }
}
